js/modules/k6/grpc/client.go

```diff
@@ -0,0 +1,574 @@
+			errc <- err
+
+			return
+		}
+		close(errc)
+	}()
+
+	if err := <-errc; err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// Invoke creates and calls a unary RPC by fully qualified method name
+//nolint: funlen,gocognit,gocyclo
+func (c *Client) Invoke(ctxPtr *context.Context,
+	method string, req goja.Value, params map[string]interface{}) (*Response, error) {
+	ctx := *ctxPtr
+	rt := common.GetRuntime(ctx)
+	state := lib.GetState(ctx)
+	if state == nil {
+		return nil, errInvokeRPCInInitContext
+	}
+
+	if c.conn == nil {
+		return nil, errors.New("no gRPC connection, you must call connect first")
+	}
+
+	if method == "" {
+		return nil, errors.New("method to invoke cannot be empty")
+	}
+
+	if method[0] != '/' {
+		method = "/" + method
+	}
+	md := c.mds[method]
+
+	if md == nil {
+		return nil, fmt.Errorf("method %q not found in file descriptors", method)
+	}
+
+	tags := state.CloneTags()
+	timeout := 60 * time.Second
+
+	ctx = metadata.NewOutgoingContext(ctx, metadata.New(nil))
+	for k, v := range params {
+		switch k {
+		case "headers":
+			rawHeaders, ok := v.(map[string]interface{})
+			if !ok {
+				return nil, errors.New("headers must be an object with key-value pairs")
+			}
+			for hk, kv := range rawHeaders {
+				// TODO(rogchap): Should we manage a string slice?
+				strVal, ok := kv.(string)
+				if !ok {
+					return nil, fmt.Errorf("header %q value must be a string", hk)
+				}
+				ctx = metadata.AppendToOutgoingContext(ctx, hk, strVal)
+			}
+		case "tags":
+			rawTags, ok := v.(map[string]interface{})
+			if !ok {
+				return nil, errors.New("tags must be an object with key-value pairs")
+			}
+			for tk, tv := range rawTags {
+				strVal, ok := tv.(string)
+				if !ok {
+					return nil, fmt.Errorf("tag %q value must be a string", tk)
+				}
+				tags[tk] = strVal
+			}
+		case "timeout":
+			var err error
+			switch t := v.(type) {
+			case string:
+				if timeout, err = time.ParseDuration(t); err != nil {
+					return nil, fmt.Errorf("unable to parse %q: %v", k, err)
+				}
+			case int64:
+				timeout = time.Duration(float64(t)) * time.Millisecond
+			case float64:
+				timeout = time.Duration(t) * time.Millisecond
+			default:
+				return nil, fmt.Errorf("unable to use type %T as a timeout value", v)
+			}
+		default:
+			return nil, fmt.Errorf("unknown param: %q", k)
+		}
+	}
+	if state.Options.SystemTags.Has(stats.TagURL) {
+		tags["url"] = fmt.Sprintf("%s%s", c.conn.Target(), method)
+	}
+
+	parts := strings.Split(method[1:], "/")
+	if state.Options.SystemTags.Has(stats.TagService) {
+		tags["service"] = parts[0]
+	}
+	if state.Options.SystemTags.Has(stats.TagMethod) {
+		tags["method"] = parts[1]
+	}
+
+	if state.Options.SystemTags.Has(stats.TagRPCType) {
+		// (rogchap) This method only supports unary RPCs
+		// if this is refactored to support streaming then this should
+		// be updated to be based on the method descriptor (IsClientStreaming/IsServerStreaming)
+		tags["rpc_type"] = "unary"
+	}
```

Comment:
Hmm I'm wondering if the `rpc_type` tag is necessary at this poing - as you say, we currently support only unary requests, so this will always be `unary`. And when we support streaming RPC calls, I'm not sure we'd be able to reuse `grpc_req_duration`, since it doesn't make a ton of sense in that context :confused: In the spirit of minimalism for the first iteration, it seems like a good idea to remove it for now. We can easily add it back later, if there are different values for the tag.

Reply:
Agreed; I added that because it was in the original ticket/issue #441 will remove.

I'm now wondering if I should remove `grpc_req_duration` and just re-use `http_req_duration`? Do you think this would be confusing for users? gRPC is still HTTP after all.

Would be nice to add some more metrics (tls handshakes, connection etc); but I've left this out to reduce the scope of this PR, vs future iterations.